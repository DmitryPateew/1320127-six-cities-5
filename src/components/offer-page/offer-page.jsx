import React, {PureComponent} from "react";
import PropTypes from "prop-types";
import {TypeOfHosing} from "../../constant";
import ReviewForm from "../review-form/review-form";
import {withRouter} from 'react-router-dom';
import {connect} from "react-redux";
import Header from "../header/header";
import {AuthorizationStatus} from "../../constant";
import {starWith} from "../../main-logic";
import {fetchCommentList, fetchNearList} from "../../store/api-actions";
import NearPlaceList from "../near-place-list/near-place-list";
import Map from "../map/map";
import {TypeOfComponent} from "../../constant";
import {sortRevievsFromFresh} from "../../main-logic";
import {formatCommentDate} from "../../main-logic";


class OfferPage extends PureComponent {
  constructor(props) {
    super(props);
    this.id = this.props.match.params.id;
  }

  componentDidMount() {
    this.props.loadNearOffers(this.id);
    this.props.loadComments(this.id);
  }

  render() {
    const {offers, authorizationStatus, comments} = this.props;
    const offerExact = offers.find((offer) => offer.id.toString() === this.id);
    const sortetComments = sortRevievsFromFresh(comments);
    return (
      <div className="page">
        <Header/>
        <main className="page__main page__main--property">
          <section className="property">
            <div className="property__gallery-container container">
              <div className="property__gallery">
                {offerExact.photos.map((photo, i) => (
                  <div key={i} className="property__image-wrapper">
                    <img className="property__image" src={photo} alt="Photo studio"/>
                  </div>
                ))}
              </div>
            </div>
            <div className="property__container container">
              <div className="property__wrapper">
                <div className="property__mark">
                  <span>{offerExact.premium === true ? `Premium` : `Usual`}</span>
                </div>
                <div className="property__name-wrapper">
                  <h1 className="property__name">
                    {offerExact.title}
                  </h1>
                  <button className="property__bookmark-button button" type="button">
                    <svg className="property__bookmark-icon" width="31" height="33">
                    </svg>
                    <span className="visually-hidden">To bookmarks</span>
                  </button>
                </div>
                <div className="property__rating rating">
                  <div className="property__stars rating__stars">
                    <span style={{width: starWith(offerExact.rating) + `%`}}></span>
                    <span className="visually-hidden">Rating</span>
                  </div>
                  <span className="property__rating-value rating__value">{offerExact.rating}</span>
                </div>
                <ul className="property__features">
                  <li className="property__feature property__feature--entire">
                    {offerExact.typeOfHousing}
                  </li>
                  <li className="property__feature property__feature--bedrooms">
                    {offerExact.countBedrooms} Bedrooms
                  </li>
                  <li className="property__feature property__feature--adults">
                    Max {offerExact.maxAdults} adults
                  </li>
                </ul>
                <div className="property__price">
                  <b className="property__price-value">&euro;{offerExact.pricePerNight}</b>
                  <span className="property__price-text">&nbsp;night</span>
                </div>
                <div className="property__inside">
                  <h2 className="property__inside-title">What&apos;s inside</h2>
                  <ul className="property__inside-list">
                    {offerExact.facilities.map((facilit, i) => (
                      <li key={i} className="property__inside-item">
                        {facilit}
                      </li>
                    ))}
                  </ul>
                </div>
                <div className="property__host">
                  <h2 className="property__host-title">Meet the host</h2>
                  <div className="property__host-user user">
                    <div className="property__avatar-wrapper property__avatar-wrapper--pro user__avatar-wrapper">
                      <img className="property__avatar user__avatar" src={offerExact.ownerInformation.photo} width="74" height="74" alt="Host avatar"/>
                    </div>
                    <span className="property__user-name">
                      {offerExact.ownerInformation.name}</span>
                    <span className="property__user-status">
                      {offerExact.ownerInformation.super === true ? `Pro` : `Usual`}</span>
                  </div>
                  <div className="property__description">
                    <p className="property__text">
                      {offerExact.description}
                    </p>
                  </div>
                </div>
                <section className="property__reviews reviews">
                  <h2 className="reviews__title">Reviews &middot; <span
                    className="reviews__amount">{comments.length ? comments.length : `0`}</span>
                  </h2>
                  <ul className="reviews__list">
                    {sortetComments.slice(0, 10).map((comment, i) => (
                      <li key={i} className="reviews__item">
                        <div className="reviews__user user">
                          <div className="reviews__avatar-wrapper user__avatar-wrapper">
                            <img className="reviews__avatar user__avatar" src={comment.user.avatar_url} width="54" height="54" alt="Reviews avatar"/>
                          </div>
                          <span className="reviews__user-name">
                            {comment.user.name}</span>
                        </div>
                        <div className="reviews__info">
                          <div className="reviews__rating rating">
                            <div className="reviews__stars rating__stars">
                              <span style={{width: starWith(comment.rating) + `%`}}></span>
                              <span className="visually-hidden">Rating{comment.rating}</span>
                            </div>
                          </div>
                          <p className="reviews__text">
                            {comment.comment}
                          </p>
                          <time className="reviews__time" dateTime="2019-04-24">{formatCommentDate(comment.date)}</time>
                        </div>
                      </li>
                    ))}
                  </ul>
                  {authorizationStatus === AuthorizationStatus.AUTH ? <ReviewForm id={this.id}/> : ``}
                </section>
              </div>
            </div>
            <section className="property__map map">
              <Map coords={offerExact.coords} type={TypeOfComponent.OFFER}/>
            </section>
          </section>
          <div className="container">
            <section className="near-places places">
              <h2 className="near-places__title">Other places in the neighbourhood</h2>
              <div className="near-places__list places__list">
                <NearPlaceList/>
              </div>
            </section>
          </div>
        </main>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  loadComments(id) {
    dispatch(fetchCommentList(id));
  },
  loadNearOffers(id) {
    dispatch(fetchNearList(id));
  },
});

const mapStateToProps = ({DATA, USER, COMMENT, NEARBY}) => ({
  data: NEARBY.nearby,
  offers: DATA.offers,
  authorizationStatus: USER.authorizationStatus,
  comments: COMMENT.comments,
});

OfferPage.propTypes = {
  match: PropTypes.any.isRequired,
  comments: PropTypes.array,
  loadNearOffers: PropTypes.func.isRequired,
  data: PropTypes.array,
  loadComments: PropTypes.func,
  authorizationStatus: PropTypes.any,
  offers: PropTypes.arrayOf(
      PropTypes.shape({
        photos: PropTypes.array.isRequired,
        premium: PropTypes.bool.isRequired,
        title: PropTypes.string.isRequired,
        rating: PropTypes.number.isRequired,
        typeOfHousing: PropTypes.oneOf([TypeOfHosing.APARTMENT, TypeOfHosing.HOTEL, TypeOfHosing.HOUSE, TypeOfHosing.ROOM]).isRequired,
        countBedrooms: PropTypes.number.isRequired,
        maxAdults: PropTypes.number.isRequired,
        pricePerNight: PropTypes.number.isRequired,
        facilities: PropTypes.array.isRequired,
        ownerInformation: PropTypes.shape({
          photo: PropTypes.string.isRequired,
          name: PropTypes.string.isRequired,
          super: PropTypes.bool.isRequired,
        }),
        description: PropTypes.string.isRequired,
        reviews: PropTypes.array,
      }
      ))
};
export {OfferPage};
export default connect(mapStateToProps, mapDispatchToProps)(withRouter(OfferPage));
