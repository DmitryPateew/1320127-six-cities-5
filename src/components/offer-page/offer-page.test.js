import React from "react";
import renderer from "react-test-renderer";
import {OfferPage} from "./offer-page";
import {expect, it} from "@jest/globals";

const offersTest = [{
  id: 1,
  city: `Hamburg`,
  coords: [52.3909553943508, 4.85309666406198],
  favorite: true,
  photos: [`img/apartment-02.jpg`, `img/apartment-01.jpg`],
  title: `Beautiful & luxurious studio at great location`,
  description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.
An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.`,
  premium: false,
  typeOfHousing: `room`,
  rating: 1,
  countBedrooms: 3,
  maxAdults: 3,
  pricePerNight: 150,
  facilities: [`Wi-Fi`, `Heating`, `Kitchen`],
}, {
  id: 2,
  city: `Hamburg`,
  coords: [52.3909553943508, 4.85309666406198],
  favorite: true,
  photos: [`img/apartment-02.jpg`, `img/apartment-01.jpg`],
  title: `Beautiful & luxurious studio at great location`,
  description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.
An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.`,
  premium: false,
  typeOfHousing: `room`,
  rating: 1,
  countBedrooms: 3,
  maxAdults: 3,
  pricePerNight: 150,
  facilities: [`Wi-Fi`, `Heating`, `Kitchen`]
}];

const testComments = [
  {
    id: 1,
    name: `Wadim`,
    rating: 4,
    date: `2019-05-08T14:13:56.569Z`,
    comment: `ffdlkmv;lfdmv;ldfm;vlmdf;`,
    user: {
      avatarUrl: `img/1.png`,
      id: 3,
      isPro: false,
      name: `Andrey`,
    }
  },
  {
    id: 2,
    name: `Wadim`,
    rating: 4,
    date: `2019-05-08T14:13:56.569Z`,
    comment: `Wadim`,
    user: {
      avatarUrl: `img/1.png`,
      id: 8,
      isPro: false,
      name: `Andrey`,
    }
  },
  {
    id: 3,
    authorPhoto: `img/avatar-max.jpg`,
    name: `Wadim`,
    rating: 4,
    date: `2019-05-08T14:13:56.569Z`,
    comment: `sdvdsvsdv`,
    user: {
      avatarUrl: `img/1.png`,
      id: 10,
      isPro: false,
      name: `Andrey`,
    }
  }
];

it(`Should OfferPage render correctly`, () => {
  const offerPage = renderer.create(<OfferPage
    loadComments={()=>{}}
    comments={testComments}
    offers={offersTest}
    authorizationStatus={`AUTH`}
  />)
    .toJSON();

  expect(offerPage).toMatchSnapshot();
});
