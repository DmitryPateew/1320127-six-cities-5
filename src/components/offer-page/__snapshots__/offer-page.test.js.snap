// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should OfferPage render correctly 1`] = `
<div
  className="page"
>
  <Connect(Header) />
  <main
    className="page__main page__main--property"
  >
    <section
      className="property"
    >
      <div
        className="property__gallery-container container"
      >
        <div
          className="property__gallery"
        >
          <div
            className="property__image-wrapper"
            key="0"
          >
            <img
              alt="Photo studio"
              className="property__image"
              src="img/apartment-02.jpg"
            />
          </div>
          <div
            className="property__image-wrapper"
            key="1"
          >
            <img
              alt="Photo studio"
              className="property__image"
              src="img/apartment-01.jpg"
            />
          </div>
        </div>
      </div>
      <div
        className="property__container container"
      >
        <div
          className="property__wrapper"
        >
          <div
            className="property__mark"
          >
            <span>
              Usual
            </span>
          </div>
          <div
            className="property__name-wrapper"
          >
            <h1
              className="property__name"
            >
              Beautiful & luxurious studio at great location
            </h1>
            <button
              className="property__bookmark-button button"
              type="button"
            >
              <svg
                className="property__bookmark-icon"
                height="33"
                width="31"
              />
              <span
                className="visually-hidden"
              >
                To bookmarks
              </span>
            </button>
          </div>
          <div
            className="property__rating rating"
          >
            <div
              className="property__stars rating__stars"
            >
              <span
                style={
                  Object {
                    "width": "20%",
                  }
                }
              />
              <span
                className="visually-hidden"
              >
                Rating
              </span>
            </div>
            <span
              className="property__rating-value rating__value"
            >
              1
            </span>
          </div>
          <ul
            className="property__features"
          >
            <li
              className="property__feature property__feature--entire"
            >
              room
            </li>
            <li
              className="property__feature property__feature--bedrooms"
            >
              3
               Bedrooms
            </li>
            <li
              className="property__feature property__feature--adults"
            >
              Max 
              3
               adults
            </li>
          </ul>
          <div
            className="property__price"
          >
            <b
              className="property__price-value"
            >
              €
              150
            </b>
            <span
              className="property__price-text"
            >
               night
            </span>
          </div>
          <div
            className="property__inside"
          >
            <h2
              className="property__inside-title"
            >
              What's inside
            </h2>
            <ul
              className="property__inside-list"
            >
              <li
                className="property__inside-item"
                key="0"
              >
                Wi-Fi
              </li>
              <li
                className="property__inside-item"
                key="1"
              >
                Heating
              </li>
              <li
                className="property__inside-item"
                key="2"
              >
                Kitchen
              </li>
            </ul>
          </div>
          <div
            className="property__host"
          >
            <h2
              className="property__host-title"
            >
              Meet the host
            </h2>
            <div
              className="property__host-user user"
            >
              <div
                className="property__avatar-wrapper property__avatar-wrapper--pro user__avatar-wrapper"
              >
                <img
                  alt="Host avatar"
                  className="property__avatar user__avatar"
                  height="74"
                  src="img/apartment-02.jpg"
                  width="74"
                />
              </div>
              <span
                className="property__user-name"
              >
                Vasja
              </span>
              <span
                className="property__user-status"
              >
                Pro
              </span>
            </div>
            <div
              className="property__description"
            >
              <p
                className="property__text"
              >
                A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.
An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.
              </p>
            </div>
          </div>
          <section
            className="property__reviews reviews"
          >
            <h2
              className="reviews__title"
            >
              Reviews · 
              <span
                className="reviews__amount"
              >
                3
              </span>
            </h2>
            <ul
              className="reviews__list"
            >
              <li
                className="reviews__item"
                key="0"
              >
                <div
                  className="reviews__user user"
                >
                  <div
                    className="reviews__avatar-wrapper user__avatar-wrapper"
                  >
                    <img
                      alt="Reviews avatar"
                      className="reviews__avatar user__avatar"
                      height="54"
                      width="54"
                    />
                  </div>
                  <span
                    className="reviews__user-name"
                  >
                    Andrey
                  </span>
                </div>
                <div
                  className="reviews__info"
                >
                  <div
                    className="reviews__rating rating"
                  >
                    <div
                      className="reviews__stars rating__stars"
                    >
                      <span
                        style={
                          Object {
                            "width": "80%",
                          }
                        }
                      />
                      <span
                        className="visually-hidden"
                      >
                        Rating
                        4
                      </span>
                    </div>
                  </div>
                  <p
                    className="reviews__text"
                  >
                    ffdlkmv;lfdmv;ldfm;vlmdf;
                  </p>
                  <time
                    className="reviews__time"
                    dateTime="2019-04-24"
                  >
                    May 2019
                  </time>
                </div>
              </li>
              <li
                className="reviews__item"
                key="1"
              >
                <div
                  className="reviews__user user"
                >
                  <div
                    className="reviews__avatar-wrapper user__avatar-wrapper"
                  >
                    <img
                      alt="Reviews avatar"
                      className="reviews__avatar user__avatar"
                      height="54"
                      width="54"
                    />
                  </div>
                  <span
                    className="reviews__user-name"
                  >
                    Andrey
                  </span>
                </div>
                <div
                  className="reviews__info"
                >
                  <div
                    className="reviews__rating rating"
                  >
                    <div
                      className="reviews__stars rating__stars"
                    >
                      <span
                        style={
                          Object {
                            "width": "80%",
                          }
                        }
                      />
                      <span
                        className="visually-hidden"
                      >
                        Rating
                        4
                      </span>
                    </div>
                  </div>
                  <p
                    className="reviews__text"
                  >
                    Wadim
                  </p>
                  <time
                    className="reviews__time"
                    dateTime="2019-04-24"
                  >
                    May 2019
                  </time>
                </div>
              </li>
              <li
                className="reviews__item"
                key="2"
              >
                <div
                  className="reviews__user user"
                >
                  <div
                    className="reviews__avatar-wrapper user__avatar-wrapper"
                  >
                    <img
                      alt="Reviews avatar"
                      className="reviews__avatar user__avatar"
                      height="54"
                      width="54"
                    />
                  </div>
                  <span
                    className="reviews__user-name"
                  >
                    Andrey
                  </span>
                </div>
                <div
                  className="reviews__info"
                >
                  <div
                    className="reviews__rating rating"
                  >
                    <div
                      className="reviews__stars rating__stars"
                    >
                      <span
                        style={
                          Object {
                            "width": "80%",
                          }
                        }
                      />
                      <span
                        className="visually-hidden"
                      >
                        Rating
                        4
                      </span>
                    </div>
                  </div>
                  <p
                    className="reviews__text"
                  >
                    sdvdsvsdv
                  </p>
                  <time
                    className="reviews__time"
                    dateTime="2019-04-24"
                  >
                    May 2019
                  </time>
                </div>
              </li>
            </ul>
            <Connect(ReviewForm)
              id="2"
            />
          </section>
        </div>
      </div>
      <section
        className="property__map map"
      >
        <Connect(Map)
          coords={
            Array [
              52.3909553943508,
              4.85309666406198,
            ]
          }
          type="OFFER"
        />
      </section>
    </section>
    <div
      className="container"
    >
      <section
        className="near-places places"
      >
        <h2
          className="near-places__title"
        >
          Other places in the neighbourhood
        </h2>
        <div
          className="near-places__list places__list"
        >
          <Connect(NearPlaceList) />
        </div>
      </section>
    </div>
  </main>
</div>
`;
